BitwardenShared/Core/Platform/Services/AuthenticatorSyncService.swift

```diff
@@ -0,0 +1,248 @@
+        self.sharedKeychainRepository = sharedKeychainRepository
+        self.notificationCenterService = notificationCenterService
+        self.stateService = stateService
+        self.vaultTimeoutService = vaultTimeoutService
+        super.init()
+    }
+
+    // MARK: Public Methods
+
+    public func start() {
+        Task {
+            if await configService.getFeatureFlag(FeatureFlag.enableAuthenticatorSync,
+                                                  defaultValue: false) {
+                subscribeToAppState()
+            }
+        }
+    }
+
+    // MARK: Private Methods
+
+    /// Check to see if the shared Authenticator key exists. If it doesn't already exist, create it.
+    ///
+    private func createAuthenticatorKeyIfNeeded() async throws {
+        let storedKey = try? await sharedKeychainRepository.getAuthenticatorKey()
+        if storedKey == nil {
+            let key = SymmetricKey(size: .bits256)
+            let data = key.withUnsafeBytes { Data(Array($0)) }
+            try await sharedKeychainRepository.setAuthenticatorKey(data)
+        }
+    }
+
+    /// Take a list of encrypted ciphers, decrypt them, filter for only active ciphers with a totp code, then
+    /// convert the list to AuthenticatorSyncItemDataModel to be stored and sync'd to the Authenticator app.
+    ///
+    /// - Parameters:
+    ///   - ciphers: The encrypted `Cipher` objects.
+    ///   - userId: The userId of the account to which these Ciphers belong.
+    ///
+    /// - Returns: The decrypted, filtered, and sorted `CipherDTO` objects.
+    ///
+    private func decryptTOTPs(_ ciphers: [Cipher],
+                              userId: String) async throws -> [AuthenticatorBridgeItemDataView] {
+        let totpCiphers = ciphers.filter { cipher in
+            cipher.deletedDate == nil
+                && cipher.type == .login
+                && cipher.login?.totp != nil
+        }
+        let decryptedCiphers = try await totpCiphers.asyncMap { cipher in
+            try await self.clientService.vault().ciphers().decrypt(cipher: cipher)
+        }
+        let account = try await stateService.getActiveAccount()
+        let username = account.profile.name ?? account.profile.email
+
+        return decryptedCiphers.map { cipher in
+            AuthenticatorBridgeItemDataView(
+                favorite: false,
+                id: cipher.id ?? UUID().uuidString,
+                name: cipher.name,
+                totpKey: cipher.login?.totp,
+                username: username
+            )
+        }
+    }
+
+    /// This function handles the initial syncing with the Authenticator app as well as listening for updates
+    /// when the user adds new items. This is called when the sync is turned on.
+    ///
+    /// - Parameter userId: The userId of the user who has turned on sync.
+    ///
+    private func handleSyncOnForUserId(_ userId: String) async {
+        Logger.application.debug("#### sync is on for userId: \(userId)")
```

Comment:
🎨  We probably don't want log statements getting merged in, especially debug ones

Reply:
Should I just pull them all out? I'm open to that. They are nice in debugging this, but we're getting much closer to just seeing the actual results directly (and I can add them in as needed while debugging).

Reply:
🤔 Which would be the downsides of having these debug logs in `main`? I mean end users won't see that, only debug builds will use that.
I often go by the rule: if the debug log will help debugging the feature at a later stage or because of some bug reported then I'd leave that and merge it but if it's something trivial or just to see where the flow goes I'd remove it from the commit.

Reply:
That's a reasonable line and I think these fall outside of it - they are useful for debugging for now, but I don't see them as being useful later when the whole system is up and running. I've pulled them all the way out in the latest commit. 👍